app/src/main/java/mozilla/lockbox/presenter/ItemDetailPresenter.kt

```diff
@@ -11,77 +11,79 @@ import io.reactivex.Observable
-                                    }
-                                }
-                                .addTo(compositeDisposable)
-                    }
-                }
-                .addTo(compositeDisposable)
+        handleClicks(view.usernameCopyClicks) {
+            val username = it.username ?: ""
+            if (username.isNotEmpty()) {
+                dispatcher.dispatch(ClipboardAction.CopyUsername(username))
+                view.showToastNotification(R.string.toast_username_copied)
+            }
+        }
 
-        this.view.passwordCopyClicks
-                .subscribe {
-                    view.itemId?.let {
-                        dataStore.get(it)
-                                .subscribe {
-                                    if (it!!.password.isNotEmpty()) {
-                                        dispatcher.dispatch(ClipboardAction.CopyPassword(it.password))
-                                        view.showToastNotification(R.string.toast_password_copied)
-                                    }
-                                }
-                                .addTo(compositeDisposable)
-                    }
-                }
-                .addTo(compositeDisposable)
+        handleClicks(view.passwordCopyClicks) {
+            if (it.password.isNotEmpty()) {
+                dispatcher.dispatch(ClipboardAction.CopyPassword(it.password))
+                view.showToastNotification(R.string.toast_password_copied)
+            }
+        }
+
+        handleClicks(view.hostnameClicks) {
+            if (it.hostname.isNotEmpty()) {
+                dispatcher.dispatch(RouteAction.OpenWebsite(it.hostname))
+            }
+        }
 
         this.view.togglePasswordClicks
-                .subscribe {
-                    view.isPasswordVisible = view.isPasswordVisible.not()
-                }
-                .addTo(compositeDisposable)
+            .subscribe {
+                view.isPasswordVisible = view.isPasswordVisible.not()
+            }
+            .addTo(compositeDisposable)
+
+        view.isPasswordVisible = false
 
         // now set up the data.
-        val itemId = view.itemId ?: return
+        val itemId = this.itemId ?: return
         dataStore.get(itemId)
-                .map {
-                    ItemDetailViewModel(it.id, titleFromHostname(it.hostname), it.hostname, it.username, it.password)
-                }
-                .subscribe(view::updateItem)
-                .addTo(compositeDisposable)
+            .map {
+                credentials = it
+                it.toDetailViewModel()
+            }
+            .subscribe(view::updateItem)
+            .addTo(compositeDisposable)
+    }
 
-        view.isPasswordVisible = false
+    private fun handleClicks(clicks: Observable<Unit>, withServerPassword: (ServerPassword) -> Unit) {
+        clicks.subscribe {
+            this.credentials?.let { withServerPassword(it) }
```

Comment:
I wonder if there's an access race condition with accessing `this.credentials?` in this way.. If the dataStore pushes an update to the item while we are handling a click, it may be getting mutated as we access it here.

Reply:
`ServerPassword` is treated as immutable, for the object instance you have won't change from underneath you.

Reply:
but the `val credentials` that we are accessing here _is_ mutable, not an instance getting passed down from another Observable.

Reply:
it gets mutated on ln. 77, which could run at the same time as this method.

Reply:
ah, now I see.  Thanks.  At least the _value_ won't change, so the problem isn't quite as deep as it could have been.

Reply:
I'm unsure there is a good answer to this. If we consider two schemes:

 1. Cache-on-display: store the `ServerPassword` in `credentials` as we give them to the view to display.
 2. No-cache: each time you use a string (for display, or copying) then get a fresh copy from the database.

For cache-on-display: 

 * If `credentials` is changed between display and user clicking copy, then the new string is both displayed and copied. 
 * If  database copy of the credential is changed between the click being detected and the copy taking place, then it will have to wait until the copy has finished, as the click, display and mutation are all on the UI thread.

For the no-cache scheme, the copy may end up copying a string newer string than that is being displayed. The user only finds this out when the user gets to the other app. 

HOWEVER: both of these scenarios are somewhat tiny considerations in comparison to the S L O W speed of the user and app switching: 

1. the user navigates to the detail screen
2. the credentials on the screen are updated
3. the user copies the old username
4. the user switches app 
5. the user pastes the old username 
6. the user switches app back to Lockbox.
6a. the credentials on the screen are updated
7. the user copies the new password 
8. the user switches app
9. the user pastes the new password.

This happens in both schemes.

Reply:
I've added an `assertOnUiThread()` method in a new file `DebugSupport`.